packages/giselle/src/engine/generations/utils.ts

```diff
@@ -30,6 +30,91 @@ import type { GenerationId } from "../../concepts/identifiers";
 import { parseAndMod } from "../../data-mod";
 import type { GiselleStorage } from "../experimental_storage";
 import type { GiselleEngineContext } from "../types";
+import type { PreparedToolSet } from "./types";
+
+const URL_CONTEXT_LIMIT = 20;
+
+export function addUrlContextTool({
+	preparedToolSet,
+	urls,
+	tool,
+}: {
+	preparedToolSet: PreparedToolSet;
+	urls: string[] | undefined;
+	tool: PreparedToolSet["toolSet"][string];
+}): PreparedToolSet {
+	if (urls === undefined || urls.length === 0) {
+		return preparedToolSet;
+	}
+
+	return {
+		...preparedToolSet,
+		toolSet: {
+			...preparedToolSet.toolSet,
+			url_context: tool,
+		},
+	};
+}
+
+type BuildMessageObjectResult = {
+	messages: ModelMessage[];
+	urlContextUrls: string[];
+};
+
+function extractHttpsUrls(text: string): string[] {
+	const matches =
+		text.match(/https?:\/\/[\w\-._~:/?#[\]@!$&'()*+,;=%]+/g) ?? [];
+	const unique = new Set<string>();
+	for (const candidate of matches) {
+		try {
+			const parsed = new URL(candidate);
+			if (parsed.protocol !== "https:") {
+				continue;
+			}
+			const normalized = parsed.toString();
+			unique.add(normalized);
+		} catch {}
+	}
+	return Array.from(unique);
+}
+
+function applyUrlContextPrompt({
```

Comment:
I don't see the need to alter the prompts entered by the user.

Reply:
Thanks for calling this out. We’ve dropped applyUrlContextPrompt entirely. The prompt now goes through untouched, just like the other tools, and URL Context simply registers the url_context tool without any client-side preprocessing. Any validation (including URL limits) is left to the API response.

[751b699](https://github.com/giselles-ai/giselle/pull/1877/commits/751b699eee0c6290e5470466488b38d59c270fa0)